### ISSUE TITLE
balance assertions drop lot annotations

### ISSUE THREAD

**Issue by taviso** [ledger/ledger#2355]
I think using balance assertions lose lot annotations, example:

```
2024/01/01 * Old Broker
    Assets:Brokerage:Old       100 FOO @ $123
    Assets:Checking
2024/01/02 * New Broker  ; Transfer
    Assets:Brokerage:Old               =0
    Assets:Brokerage:New
```

```
$ ledger --args-only --file test.ldg bal --lots ^Assets:Brokerage:New
             100 FOO  Assets:Brokerage:New
```

It looks like it dropped the cost basis and date information. I think this is a bug.

**Comment by taviso** [comment]
Right, because the lot annotations were dropped.

**Comment by taviso** [comment]
Hmm thanks, that does look like it. It looks like the intention was to allow `= 100 FOO` to work even with lot annotations, I guess that makes sense, but it breaks my use case and your use case :(

Maybe the solution is to add non-lot-stripped assertions back, but with a slightly different syntax?

Maybe  `== 100 FOO` because `==` is more equal than `=`? 

Maybe we can get @jwiegley input 😄

**Comment by jwiegley** [comment]
It sounds like I should revert ce8fec9 and then seek a different solution to whatever prompted me to include it...

**Comment by tbm** [comment]
@taviso @riastradh can one of you submit a PR to add testcases for this and the other ticket, so this won't accidentally get broken again.  (If not, let me know and I can probably add it some time.)

**Comment by taviso** [comment]
Yep no problem, I think the solution isn't known yet... I can attempt a patch, I was just waiting for John's thoughts on if a new syntax is okay, or something else!

**Comment by jwiegley** [comment]
Actually, I wouldn't mean `== 0`, but I wouldn't change the meaning in the way described above. Here is what it would mean:

    ACCOUNT  = X

This sets the balance of `ACCOUNT` to `X`, as it does now. If `X` has no lot information, it should set the balance to an amount with no lot details, since that is what the user specified. Whether this zeroes out everything that does have lot details, or if it only affects lots in that currency without details, is something we need to decide. I think the present behavior is that it should only affects lots with no details if it has no details, or lots with details if it does have details.

    ACCOUNT  == X

This does not change the balance of `ACCOUNT` at all, but ensures it equals `X`. If there are no lot details specified then it could go either way: matching only against lots with no details, or acting as if `--lots` was not being used at all. I don't want the behavior of the assertion to depend on a command-line flag, however.

So I think the underlying design issue here is the meaning of the amount to the right of the equals sign in either case: What does it mean if it does or does not have lot details, with respect to the lots in the account under consideration?

**Comment by tbm** [comment]
> Yep no problem, I think the solution isn't known yet

Your test cases work with the old (and now restored) behaviour, so imho you can add the test cases.  If new syntax is introduced which changes behaviour, we can always update test cases along with that change.  But it's up to @jwiegley what he prefers.
